test/legacy_test/test_lr_scheduler.py

```diff
@@ -214,6 +214,203 @@ def _test_dygraph(self, place, kwargs):
         self.assertEqual(scheduler.last_lr, scheduler1.last_lr)
 
 
+def cosine_annealing_warm_restarts_lr(epoch_num, v_l):
+    if epoch_num is None and v_l['last_epoch'] < 0:
+        epoch_num = 0
+
+    cur_lr = (
+        v_l['eta_min']
+        + (v_l['base_lr'] - v_l['eta_min'])
+        * (1 + math.cos(math.pi * v_l['T_cur'] / v_l['T_i']))
+        / 2
+    )
+
+    if v_l['last_epoch'] == -1:
+        cur_lr = v_l['base_lr']
+
+    if epoch_num is None:
+        epoch_num = v_l['last_epoch'] + 1
+        v_l['T_cur'] = v_l['T_cur'] + 1
+        if v_l['T_cur'] >= v_l['T_i']:
+            v_l['T_cur'] = v_l['T_cur'] - v_l['T_i']
+            v_l['T_i'] = v_l['T_i'] * v_l['T_mult']
+    else:
+        if epoch_num < 0:
+            raise ValueError(
+                f"Expected non-negative epoch, but got {epoch_num}"
+            )
+        if epoch_num >= v_l['T_0']:
+            if v_l['T_mult'] == 1:
+                v_l['T_cur'] = epoch_num % v_l['T_0']
+            else:
+                n = int(
+                    math.log(
+                        (epoch_num / v_l['T_0'] * (v_l['T_mult'] - 1) + 1),
+                        v_l['T_mult'],
+                    )
+                )
+                v_l['T_cur'] = epoch_num - v_l['T_0'] * (
+                    v_l['T_mult'] ** n - 1
+                ) / (v_l['T_mult'] - 1)
+                v_l['T_i'] = v_l['T_0'] * v_l['T_mult'] ** (n)
+        else:
+            v_l['T_i'] = v_l['T_0']
+            v_l['T_cur'] = epoch_num
+    v_l['last_epoch'] = math.floor(epoch_num)
+
+    return cur_lr
+
+
+class TestCosineAnnealingWarmRestarts(unittest.TestCase):
+    def test_CosineRestartsLR(self):
+        # check value of T_0
+        with self.assertRaises(ValueError):
+            paddle.optimizer.lr.CosineAnnealingWarmRestarts(
+                learning_rate=0.5,
+                T_0=-1,
+                T_mult=1,
+            )
+        # check type of T_0
+        with self.assertRaises(ValueError):
+            paddle.optimizer.lr.CosineAnnealingWarmRestarts(
+                learning_rate=0.5,
+                T_0=1.0,
+                T_mult=1,
+            )
+        # check value of T_mult
+        with self.assertRaises(ValueError):
+            paddle.optimizer.lr.CosineAnnealingWarmRestarts(
+                learning_rate=0.5,
+                T_0=1,
+                T_mult=-1,
+            )
+        # check type of T_mult
+        with self.assertRaises(ValueError):
+            paddle.optimizer.lr.CosineAnnealingWarmRestarts(
+                learning_rate=0.5,
+                T_0=1,
+                T_mult=1.0,
+            )
+
+        places = [paddle.CPUPlace()]
+        if core.is_compiled_with_cuda():
+            places.append(paddle.CUDAPlace(0))
+
+        for place in places:
+            kwargs = {
+                'learning_rate': 0.5,
+                'T_0': 1,
```

Comment:
`T_0 = 1` is special case, should add more test case when `T_0 > 1`

Reply:
Done